WooCommerce/Classes/ViewModels/Order Details/Shipping Labels/ShippingLabelOrderItemsAggregator.swift

```diff
@@ -0,0 +1,153 @@
+    ///     the product/variation has changed after order creation.
+    ///   - products: An array of products for an order that have been fetched.
+    ///   - productVariations: An array of product variations for an order that have been fetched.
+    ///   - currencyFormatter: Used to convert a product/variation's price string to number.
+    init(shippingLabels: [ShippingLabel],
+         orderItems: [OrderItem],
+         products: [Product],
+         productVariations: [ProductVariation],
+         currencyFormatter: CurrencyFormatter = CurrencyFormatter(currencySettings: ServiceLocator.currencySettings)) {
+        self.currencyFormatter = currencyFormatter
+
+        aggregateProductsToOrderItems(shippingLabels: shippingLabels, orderItems: orderItems, products: products, productVariations: productVariations)
+    }
+
+    /// Returns an array of order items for a shipping label.
+    func orderItems(of shippingLabel: ShippingLabel) -> [AggregateOrderItem] {
+        orderItemsByShippingLabelID[shippingLabel.shippingLabelID] ?? []
+    }
+}
+
+private extension ShippingLabelOrderItemsAggregator {
+    mutating func aggregateProductsToOrderItems(shippingLabels: [ShippingLabel],
+                                                orderItems: [OrderItem],
+                                                products: [Product],
+                                                productVariations: [ProductVariation]) {
+        shippingLabels.forEach { aggregateProductsToOrderItems(shippingLabel: $0,
+                                                               orderItems: orderItems,
+                                                               products: products,
+                                                               productVariations: productVariations) }
+    }
+
+    mutating func aggregateProductsToOrderItems(shippingLabel: ShippingLabel,
+                                                orderItems: [OrderItem],
+                                                products: [Product],
+                                                productVariations: [ProductVariation]) {
+        // ShippingLabel's `productNames` is always available, but `productIDs` is only available in WooCommerce Shipping & Tax v1.24.1+.
+        // Here we map a ShippingLabel's `productNames` to `ProductInformation` with an optional product ID at the corresponding index.
+        let productInfoArray = shippingLabel.productNames.enumerated().map { index, productName in
+            ProductInformation(id: shippingLabel.productIDs[safe: index], name: productName)
+        }
+        // In the API, if a product has quantity higher than 1, it is repeated in the `productNames`/`productIDs` array.
+        // Note that decimal quantity might have unexpected behavior given the current API design.
+        let uniqueProductInfoArray = productInfoArray.removingDuplicates()
```

Comment:
ℹ️ Maybe using `Set` is better so we don't rely on our own implementation. 😄 Hopefully, the Swift developers used a better algorithm. 

```suggestion
        let uniqueProductInfoArray = Array(Set(productInfoArray))
```

Reply:
There's actually a potential difference in the output using `Array(Set(array))` vs. `array.removingDuplicates()` - the former could output the unique elements in any order from the conversion to `Set`, while the latter keeps the original order and removes duplicates after the first instance. For a shipping label's order items, I think we'd like to keep the order as in the API response (product IDs/names). Feel free to correct anything!

Reply:
You're right! I stand corrected. 🤸